File: internal/replay/replay.go
// Package replay orchestrates sequential replay against both stacks.
package replay

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/pajamasi726/mocking-box/internal/binlog"
	"github.com/pajamasi726/mocking-box/internal/config"
	"github.com/pajamasi726/mocking-box/internal/corpus"
	"github.com/pajamasi726/mocking-box/internal/diff"
)

const maxBodyBytes = 1 << 20 // 1 MiB per response kept for the report

// Result is the outcome of replaying one request against both stacks.
type Result struct {
	Name        string            `json:"name"`
	Request     string            `json:"request"`
	Verdict     string            `json:"verdict"`
	Differences []diff.Difference `json:"differences"`
	OldStatus   int               `json:"old_status,omitempty"`
	NewStatus   int               `json:"new_status,omitempty"`
	OldBody     string            `json:"old_body,omitempty"`
	NewBody     string            `json:"new_body,omitempty"`
	OldWriteset []diff.WriteEntry `json:"old_writeset"`
	NewWriteset []diff.WriteEntry `json:"new_writeset"`
	Error       string            `json:"error,omitempty"`
	DurationMs  int64             `json:"duration_ms"`
}

type stackRuntime struct {
	cfg     config.Stack
	capture *binlog.Capture
}

type Runner struct {
	cfg    *config.Config
	client *http.Client
	old    *stackRuntime
	new    *stackRuntime

	// OnProgress, when set, is called before each request is replayed.
	OnProgress func(done, total int, name string)
}

func NewRunner(cfg *config.Config) *Runner {
	newRt := func(stack config.Stack, serverID uint32) *stackRuntime {
		rt := &stackRuntime{cfg: stack}
		if stack.MySQL != nil {
			rt.capture = binlog.New(stack.Name, stack.MySQL, serverID)
		}
		return rt
	}
	return &Runner{
		cfg:    cfg,
		client: &http.Client{Timeout: time.Duration(cfg.HTTPTimeoutS * float64(time.Second))},
		old:    newRt(cfg.Old, 5501),
		new:    newRt(cfg.New, 5502),
	}
}

func (r *Runner) Start() error {
	for _, rt := range []*stackRuntime{r.old, r.new} {
		if rt.capture != nil {
			if err := rt.capture.Start(); err != nil {
				return err
			}
		}
	}
	return nil
}

func (r *Runner) Stop() {
	for _, rt := range []*stackRuntime{r.old, r.new} {
		if rt.capture != nil {
			rt.capture.Stop()
		}
	}
}

func (r *Runner) Run(specs []corpus.RequestSpec) []Result {
	results := make([]Result, 0, len(specs))
	for i, spec := range specs {
		log.Printf("(%d/%d) %s  %s", i+1, len(specs), spec.Name, spec.Describe())
		if r.OnProgress != nil {
			r.OnProgress(i, len(specs), spec.Name)
		}
		results = append(results, r.runOne(spec))
	}
	if r.OnProgress != nil {
		r.OnProgress(len(specs), len(specs), "")
	}
	return results
}

type fired struct {
	status  int
	body    string
	headers map[string]string
	changes []binlog.RowChange
}

func (r *Runner) runOne(spec corpus.RequestSpec) Result {
	started := time.Now()
	result := Result{Name: spec.Name, Request: spec.Describe(), Verdict: diff.Error}
	defer func() { result.DurationMs = time.Since(started).Milliseconds() }()

	oldRes, err := r.fire(r.old, spec)
	if err != nil {
		result.Error = fmt.Sprintf("old: %v", err)
		return result
	}
	newRes, err := r.fire(r.new, spec)
	if err != nil {
		result.Error = fmt.Sprintf("new: %v", err)
		return result
	}

	result.OldStatus, result.NewStatus = oldRes.status, newRes.status
	result.OldBody, result.NewBody = oldRes.body, newRes.body

	sortRules := make([]diff.SortRule, len(r.cfg.Compare.SortArrays))
	for i, sr := range r.cfg.Compare.SortArrays {
		sortRules[i] = diff.SortRule{Path: sr.Path, By: sr.By}
	}
	responseDiffs := diff.DiffResponses(
		oldRes.status, oldRes.body,
		newRes.status, newRes.body,
		oldRes.headers, newRes.headers,
		diff.Options{
			NoisePaths:     r.cfg.Noise.ResponsePaths,
			CompareHeaders: r.cfg.CompareHeaders,
			SortArrays:     sortRules,
		},
	)

	result.OldWriteset = orEmptyWS(diff.NormalizeWriteset(oldRes.changes, r.cfg.Noise.Columns, r.cfg.Noise.TablesIgnore))
	result.NewWriteset = orEmptyWS(diff.NormalizeWriteset(newRes.changes, r.cfg.Noise.Columns, r.cfg.Noise.TablesIgnore))
	writesetDiffs := diff.DiffWritesets(result.OldWriteset, result.NewWriteset)

	result.Differences = append(responseDiffs, writesetDiffs...)
	result.Verdict = diff.VerdictOf(responseDiffs, writesetDiffs)
	return result
}

func (r *Runner) fire(rt *stackRuntime, spec corpus.RequestSpec) (*fired, error) {
	if rt.capture != nil {
		rt.capture.BeginWindow()
	}

	bodyBytes, isJSON, err := spec.BodyBytes()
	if err != nil {
		return nil, err
	}
	var bodyReader io.Reader
	if bodyBytes != nil {
		bodyReader = bytes.NewReader(bodyBytes)
	}
	req, err := http.NewRequest(spec.Method, rt.cfg.BaseURL+spec.Path, bodyReader)
	if err != nil {
		return nil, err
	}
	for k, v := range spec.Headers {
		req.Header.Set(k, v)
	}
	if isJSON && req.Header.Get("content-type") == "" {
		req.Header.Set("content-type", "application/json")
	}

	resp, err := r.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(io.LimitReader(resp.Body, maxBodyBytes))
	if err != nil {
		return nil, err
	}

	headers := map[string]string{}
	for k := range resp.Header {
		headers[toLower(k)] = resp.Header.Get(k)
	}

	out := &fired{status: resp.StatusCode, body: string(body), headers: headers}
	if rt.capture != nil {
		out.changes, err = rt.capture.TakeWindow(r.cfg.Attribution)
		if err != nil {
			return nil, err
		}
	}
	return out, nil
}

// orEmptyWS keeps empty write-sets as [] (not null) in report JSON.
func orEmptyWS(ws []diff.WriteEntry) []diff.WriteEntry {
	if ws == nil {
		return []diff.WriteEntry{}
	}
	return ws
}

func toLower(s string) string {
	b := []byte(s)
	for i, c := range b {
		if c >= 'A' && c <= 'Z' {
			b[i] = c + 32
		}
	}
	return string(b)
}

File: poc/python/mockingbox/replay.py
"""Sequential replay orchestrator.

For each request in the corpus:
    1. open attribution window on the old stack's binlog, send request, quiesce
    2. same against the new stack
    3. diff responses + write-sets -> verdict
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass

import httpx

from .binlog import BinlogCapture
from .config import Config, StackConfig
from .corpus import RequestSpec
from .diff import (
    ERROR,
    RequestResult,
    diff_responses,
    diff_writesets,
    normalize_writeset,
    verdict_of,
)

log = logging.getLogger(__name__)


@dataclass
class StackRuntime:
    config: StackConfig
    client: httpx.Client
    capture: BinlogCapture | None


class Replayer:
    def __init__(self, config: Config):
        self.config = config
        self.old = self._runtime(config.old, server_id=5501)
        self.new = self._runtime(config.new, server_id=5502)

    def _runtime(self, stack: StackConfig, server_id: int) -> StackRuntime:
        capture = None
        if stack.mysql is not None:
            capture = BinlogCapture(stack.name, stack.mysql, server_id=server_id)
        client = httpx.Client(
            base_url=stack.base_url, timeout=self.config.http_timeout_s
        )
        return StackRuntime(stack, client, capture)

    def start(self) -> None:
        for rt in (self.old, self.new):
            if rt.capture:
                rt.capture.start()

    def stop(self) -> None:
        for rt in (self.old, self.new):
            if rt.capture:
                rt.capture.stop()
            rt.client.close()

    # ------------------------------------------------------------------

    def run(self, corpus: list[RequestSpec]) -> list[RequestResult]:
        results = []
        total = len(corpus)
        for i, spec in enumerate(corpus, start=1):
            log.info("(%d/%d) %s  %s", i, total, spec.name, spec.describe())
            results.append(self._run_one(spec))
        return results

    def _run_one(self, spec: RequestSpec) -> RequestResult:
        result = RequestResult(name=spec.name, request=spec.describe(), verdict=ERROR)
        try:
            old_resp, old_changes = self._fire(self.old, spec)
            new_resp, new_changes = self._fire(self.new, spec)
        except Exception as exc:  # noqa: BLE001 - recorded per-request, replay continues
            log.error("  %s: %s", type(exc).__name__, exc)
            result.error = f"{type(exc).__name__}: {exc}"
            return result

        result.old_status, result.new_status = old_resp.status_code, new_resp.status_code
        response_diffs = diff_responses(
            old_resp.status_code,
            old_resp.text,
            new_resp.status_code,
            new_resp.text,
            self.config.noise.response_paths,
            dict(old_resp.headers),
            dict(new_resp.headers),
            self.config.compare_headers,
        )

        old_ws = normalize_writeset(
            old_changes, self.config.noise.columns, self.config.noise.tables_ignore
        )
        new_ws = normalize_writeset(
            new_changes, self.config.noise.columns, self.config.noise.tables_ignore
        )
        result.old_writes, result.new_writes = len(old_ws), len(new_ws)
        result.old_writeset, result.new_writeset = old_ws, new_ws
        writeset_diffs = diff_writesets(old_ws, new_ws)

        result.differences = response_diffs + writeset_diffs
        result.verdict = verdict_of(response_diffs, writeset_diffs)
        return result

    def _fire(self, rt: StackRuntime, spec: RequestSpec):
        if rt.capture:
            rt.capture.begin_window()

        kwargs: dict = {"headers": spec.headers}
        if isinstance(spec.body, (dict, list)):
            kwargs["content"] = json.dumps(spec.body)
            kwargs["headers"] = {"content-type": "application/json", **spec.headers}
        elif isinstance(spec.body, str):
            kwargs["content"] = spec.body

        response = rt.client.request(spec.method, spec.path, **kwargs)

        changes = (
            rt.capture.take_window(self.config.attribution) if rt.capture else []
        )
        return response, changes
